testing/test_termination.py

```diff
@@ -97,8 +97,12 @@ def test_close_initiating_remote_no_error(testdir, anypython):
 
 
 def test_terminate_implicit_does_trykill(testdir, anypython, capfd, pool):
-    if pool.execmodel != "thread":
+    if pool.execmodel.backend not in ("thread", "main_thread_only"):
         pytest.xfail("only os threading model supported")
+    if sys.version_info >= (3, 12):
+        pytest.xfail(
+            "since python3.12 this test triggers RuntimeError: can't create new thread at interpreter shutdown"
```

Comment:
For this safe_terminate python3.12 incompatibility with spawning threads during interpreter shutdown, a possible solution is for Group to implement the context manager protocol and call terminate in the `__exit__`  method. Explicit termination at the tail of the test script suppresses the issue because all of the gateways are already terminated when the atexit hook finally executes:

```diff
--- a/testing/test_termination.py
+++ b/testing/test_termination.py
@@ -123,4 +123,5 @@ def test_terminate_implicit_does_trykill(testdir, anypython, capfd, pool):
         # see Issue #5319 (from the release notes of 3.2 Alpha 2)
         sys.stdout = FlushNoOp()
+        group.terminate()
 
         #  use process at-exit group.terminate call
```
Maybe `__del__` is a viable alternative though, since the test also passes under python3.12 with this alternative change:
```python
--- a/src/execnet/multi.py
+++ b/src/execnet/multi.py
@@ -42 +42,3 @@ class Group:
-        atexit.register(self._cleanup_atexit)
+
+    def __del__(self):
+        self._cleanup_atexit()
```

Reply:
lets leave that for later, we ought not to use __del__

we might get a good mileage once the internals are sans-io and the "receiver thread" can use asyncio instead of the current stuff whose architecture dates back to python 2.4 in back in 2004-2005 when this library started as std.execnet as part of  pypy